### PR TITLE
Use Brigade for coverage

### PR DESCRIPTION
**Pull request by meyerbaptiste**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | I hope
| Fixed tickets | #1489
| License       | MIT
| Doc PR        | N/A

### CODE REVIEW

File: brigade.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    const job = new Job('coverage', 'php:rc-alpine');
+
+    job.tasks = [
+        'set -xe',
+        'apk add --no-cache curl git openssl icu-libs zlib icu-dev zlib-dev $PHPIZE_DEPS',
+        'docker-php-ext-install intl zip',
+        'export COMPOSER_ALLOW_SUPERUSER=1',
+        'curl -s -f -L -o /tmp/installer.php https://getcomposer.org/installer',
+        'curl -s -f -L -o /tmp/install.sig https://composer.github.io/installer.sig',
+        'php -r "\$signature = trim(file_get_contents(\'/tmp/install.sig\')); \$hash = hash(\'SHA384\', file_get_contents(\'/tmp/installer.php\')); if (!hash_equals(\$signature, \$hash)) { unlink(\'/tmp/installer.php\'); echo \'Integrity check failed, installer is either corrupt or worse.\' . PHP_EOL; exit(1); }"',
```
Comment (soyuka) [inline brigade.js]: Can't we use template strings? Would keep us from escaping the `'` and allow multi lines on complex tasks like this one.

**Comment by alanpoulain** [comment]
Shouldn't we handle the versions used (for php-code-coverage or coveralls for example) elsewhere?

File: brigade.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+const { events, Job } = require('brigadier');
+
+const coverage = (event, project) => {
+    console.log(`===> Building ${ project.repo.cloneURL } ${ event.commit }`);
+
+    const job = new Job('coverage', 'php:rc-alpine');
```
Comment (dunglas) [inline brigade.js]: Maybe can you start from the `composer` image directly ?

Comment (soyuka) [reply brigade.js]: :+1:  Like I did in the DockerFile: https://gist.github.com/soyuka/6aa9da5649907372279a88e4eb0555ef

**Comment by dunglas** [comment]
@alanpoulain, in the end, we should probably use a pre-built Docker image containing all those tools to speed up the builds.

**Comment by meyerbaptiste** [comment]
Anyway, it doesn't work for now 😢

**Comment by meyerbaptiste** [comment]
It works!

**Comment by Simperfit** [comment]
I'm glad that this is working, congrats @meyerbaptiste !

File: brigade.js (hunk 2 above)
Comment (dunglas) [reply brigade.js]: We should use a custom Docker image stored on the `coopTilleuls` account to save some CPU and speed up the builds.

**Comment by dunglas** [comment]
Wouhou!

File: brigade.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    job.tasks = [
+        `set -xe`,
```
Comment (Gregcop1) [inline brigade.js]: it's ok to use backticks for multiple lines but you can keep simple quotes here

**Comment by sroze** [comment]
But... the brigade commit status is a 404 URL and has Kevin's face on it 😄

**Comment by sroze** [comment]
What is the point of this by the way? 🤔  (Except the fancyness that I like 😛)

**Comment by dunglas** [comment]
> What is the point of this by the way? 🤔 (Except the fancyness that I like 😛)

We cannot compute the code coverage on Travis anymore (we consume too much resources). So we use Brigade to compute the coverage.

**Comment by Gregcop1** [comment]
A brigade with @dunglas as avatar is not something you see every day :p

**Comment by soyuka** [comment]
> What is the point of this by the way? :thinking: (Except the fancyness that I like :stuck_out_tongue:)

The full story: https://gist.github.com/soyuka/2c6d55da59b1ab74a9a92deed093df02

**Comment by meyerbaptiste** [comment]
@sroze:

> the brigade commit status is a 404 URL

See https://github.com/Azure/brigade/issues/126.
Another problematic issue is https://github.com/Azure/brigade/issues/125.

> has Kevin's face on it

Yes it sounds weird but it's because the Github token used comes from @dunglas' account. Maybe we should create a specific account for API Platform?

File: brigade.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    const job = new Job('coverage', 'cooptilleuls/api-platform-core-brigade-coverage:latest');
+
+    job.tasks = [
+        'set -xe',
+        'cd /src/',
+        'mkdir -p build/logs build/cov',
+        'composer require --dev --no-update \'phpunit/php-code-coverage:^5.2.2\'',
+        'composer update --prefer-dist --no-progress --no-suggest --ansi',
+        'phpdbg -qrr vendor/bin/phpunit --coverage-php build/cov/coverage-phpunit.cov',
+        'for f in $(find features -name \'*.feature\'); do FEATURE=${f//\\//_} phpdbg -qrr vendor/bin/behat --format=progress --profile coverage $f || exit $?; done',
```
Comment (soyuka) [inline brigade.js]: here literals

File: brigade.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+const coverage = (event, project) => {
+    console.log(`===> Building ${ project.repo.cloneURL } ${ event.commit }`);
+
+    const job = new Job('coverage', 'cooptilleuls/api-platform-core-brigade-coverage:latest');
+
+    job.tasks = [
+        'set -xe',
+        'cd /src/',
+        'mkdir -p build/logs build/cov',
+        'composer require --dev --no-update \'phpunit/php-code-coverage:^5.2.2\'',
```
Comment (soyuka) [inline brigade.js]: here literals

File: brigade.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        'phpdbg -qrr vendor/bin/phpunit --coverage-php build/cov/coverage-phpunit.cov',
+        'for f in $(find features -name \'*.feature\'); do FEATURE=${f//\\//_} phpdbg -qrr vendor/bin/behat --format=progress --profile coverage $f || exit $?; done',
+        'phpdbg -qrr $(which phpcov) merge --clover build/logs/clover.xml build/cov;',
+        'coveralls -v',
+    ];
+
+    job.env = {
+        'COVERALLS_RUN_LOCALLY': '1',
+        'COVERALLS_REPO_TOKEN': project.secrets.coverallsToken,
+    };
```
Comment (soyuka) [inline brigade.js]: quotes are not needed on keys

File: brigade.js (hunk 2 above)
Comment (meyerbaptiste) [reply brigade.js]: Done @dunglas. Can you please review?

**Comment by soyuka** [comment]
:fireworks: thanks @meyerbaptiste!

**Comment by sroze** [comment]
The 404 link is expected?

**Comment by meyerbaptiste** [comment]
@sroze: https://github.com/api-platform/core/pull/1505#issuecomment-346319454 😛

**Comment by sroze** [comment]
🙊 🏃

**Comment by soyuka** [comment]
@sroze anyway not a problem because this is only to keep the coverage up to date, we don't really care about the job itself :).

**Pull request merged**